### PR TITLE
SUBMARINE-558. Define Swagger API for pre-defined template submission

### PR DESCRIPTION
**Pull request by JohnTing**
### What is this PR for?
Make basic rest api for submi template.

Convert submitted template to experiment
post
http://localhost/V1/template/submit

### What type of PR is it?
[Bug Fix | Improvement | Feature | Documentation | Hot Fix | Refactoring]

### Todos
* [x] - API
* [x] - test

### What is the Jira issue?
https://issues.apache.org/jira/projects/SUBMARINE/issues/SUBMARINE-558

### How should this be tested?

```sh
# Register experiment template
curl -X POST -H "Content-Type: application/json" -d '
{
  "name": "tf-mnist-test",
  "author": "author",
  "description": "This is a template to run tf-mnist\n",
  "parameters": [
    {
      "name": "training.learning_rate",
      "value": 0.1,
      "required": true,
      "description": " mnist learning_rate "
    },
    {
      "name": "training.batch_size",
      "value": 150,
      "required": false,
      "description": "This is batch size of training"
    }
  ],
  "experimentSpec": {
    "meta": {
      "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate={{training.learning_rate}} --batch_size={{training.batch_size}}",
      "name": "tf-mnist-template-test",
      "envVars": {
        "ENV1": "ENV1"
      },
      "framework": "TensorFlow",
      "namespace": "default"
    },
    "spec": {
      "Ps": {
        "replicas": 1,
        "resources": "cpu=1,memory=1024M"
      },
      "Worker": {
        "replicas": 1,
        "resources": "cpu=1,memory=1024M"
      }
    },
    "environment": {
      "image": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
    }
  }
}
' http://127.0.0.1:8080/api/v1/template
```

```sh
# Submit experiment template to experiment 
curl -X POST -H "Content-Type: application/json" -d '
{
    "name": "tf-mnist-test", 
    "params": {
        "training.learning_rate":"0.01", 
        "training.batch_size":"150"
    }
}
' http://127.0.0.1:8080/api/v1/template/submit
```

If the submission is successful, it will return
```json
{
    "status": "OK",
    "code": 200,
    "success": true,
    "message": null,
    "result": {
        "experimentId": "experiment_1597853926000_0035",
        "name": "tf-mnist-template-test",
        "uid": "0a79d641-871e-4ba6-9e3e-eab1d4690f4e",
        "status": "Accepted",
        "acceptedTime": "2020-08-22T18:36:32.000+08:00",
        "createdTime": null,
        "runningTime": null,
        "finishedTime": null,
        "spec": {
            "meta": {
                "name": "tf-mnist-template-test",
                "namespace": "default",
                "framework": "TensorFlow",
                "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate=0.01 --batch_size=150",
                "envVars": {
                    "ENV1": "ENV1"
                }
            },
            "environment": {
                "name": null,
                "dockerImage": null,
                "kernelSpec": null,
                "description": null,
                "image": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
            },
            "spec": {
                "Ps": {
                    "replicas": 1,
                    "resources": "cpu=1,memory=1024M",
                    "name": null,
                    "image": null,
                    "cmd": null,
                    "envVars": null,
                    "resourceMap": {
                        "memory": "1024M",
                        "cpu": "1"
                    }
                },
                "Worker": {
                    "replicas": 1,
                    "resources": "cpu=1,memory=1024M",
                    "name": null,
                    "image": null,
                    "cmd": null,
                    "envVars": null,
                    "resourceMap": {
                        "memory": "1024M",
                        "cpu": "1"
                    }
                }
            }
        }
    },
    "attributes": {}
}
```

### Screenshots (if appropriate)
image

### Questions:
* Does the licenses files need update? Yes/No
* Is there breaking changes for older versions? Yes/No
* Does this needs documentation? Yes/No

### CODE REVIEW

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/experimenttemplate/ExperimentTemplateManager.java [hunk 1]
```diff
@@ -297,27 +338,27 @@ private ExperimentTemplateSpec parameterMapping(String spec) {
...
-            "Parameters contains unused key: " + parmMap.keySet());
+            "Parameters contains unused key: " + paramMap.keySet());
     }
 
     if (unmappedKeys.size() > 0) {
       throw new SubmarineRuntimeException(Status.BAD_REQUEST.getStatusCode(),
           "Template contains unmapped key: " + unmappedKeys);
     }  
-
+    ExperimentTemplateSpec tplSpec;
```
Comment (pingsutw) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/experimenttemplate/ExperimentTemplateManager.java:361]:
```suggestion
    ExperimentTemplateSpec experimentTemplateSpec;

```

**Review by pingsutw — COMMENTED** [review]
Thanks @JohnTing for working on this.
Some suggestion in the below.
1. Could we remove `test_template_2.json`? it's same as `test_template_1.json`
2. Need to change directory name `submarine-server/server-core/src/test/resources/experimenttemplate/test_template_1.json` to `submarine-server/server-core/src/test/resources/experimentTemplate/test_template_1.json`

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/experimenttemplate/ExperimentTemplateManager.java [hunk 2]
```diff
@@ -276,19 +280,56 @@ private ExperimentTemplate getExperimentTemplateDetails(String name) throws Subm
...
+
+    if (SubmittedParam == null) {
+      throw new SubmarineRuntimeException(Status.BAD_REQUEST.getStatusCode(), 
+            "Invalid ExperimentTemplateSubmit spec.");
+    }
+
+    ExperimentTemplate experimentTemplate = getExperimentTemplate(SubmittedParam.getName());
+    Map<String, String> sparam = SubmittedParam.getParams();
+
+    for (ExperimentTemplateParamSpec tpaam: experimentTemplate.getExperimentTemplateSpec().getParameters()) {
```
[the remark this answers is not in the document]
Comment (JohnTing) [reply submarine-server/server-core/src/main/java/org/apache/submarine/server/experimenttemplate/ExperimentTemplateManager.java]: This change needs to make the following json format 

```json
"experimentTemplateSpec": {
    "name": "tf-mnist-test",
    "author": "author",
    "description": "This is a template to run tf-mnist\n",
    "parameters": [
        {
            "name": "training.learning_rate",
            "required": "true",
            "description": " mnist learning_rate ",
            "value": "0.01"
        },
        {
            "name": "training.batch_size",
            "required": "false",
            "description": "This is batch size of training",
            "value": "150"
        }
    ],
}
```
into this

```json
"experimentTemplateSpec": {
    "name": "tf-mnist-test",
    "author": "author",
    "description": "This is a template to run tf-mnist\n",
    "experimentTemplateParamSpec": [
        {
            "name": "training.learning_rate",
            "required": "true",
            "description": " mnist learning_rate ",
            "value": "0.01"
        },
        {
            "name": "training.batch_size",
            "required": "false",
            "description": "This is batch size of training",
            "value": "150"
        }
    ],
}
```

I am not sure if it is appropriate.

**Comment by wangdatan** [comment]
Thanks @JohnTing for the patch, I reviewed registration flow, and what does the API looks like. I actually like the new implementation more than the original design of https://github.com/apache/submarine/blob/master/docs/design/experiment-implementation.md. The parameterized implementation is very clean and very easy for users to register and understand. Great job! 

I will leave the detailed code review works to others, at a high-level, the implementation looks good!

**Review by jiwq — COMMENTED** [review]
@JohnTing Thanks for your contribution. But I have a doubt "how can the user submit the experiment by template?" I saw the REST api `/api/v1/template/submit`, but I can't get its meaning. Such as 'create new template' or others? In my opinion, if we provide the template resource, the experiment resource depends on template. So we should extend the experiment api to support accepts the template id and related params to training.

**Comment by JohnTing** [comment]
> @JohnTing Thanks for your contribution. But I have a doubt "how can the user submit the experiment by template?" I saw the REST api `/api/v1/template/submit`, but I can't get its meaning. Such as 'create new template' or others? In my opinion, if we provide the template resource, the experiment resource depends on template. So we should extend the experiment api to support accepts the template id and related params to training.



> @JohnTing Thanks for your contribution. But I have a doubt "how can the user submit the experiment by template?" I saw the REST api `/api/v1/template/submit`, but I can't get its meaning. Such as 'create new template' or others? In my opinion, if we provide the template resource, the experiment resource depends on template. So we should extend the experiment api to support accepts the template id and related params to training.



> @JohnTing Thanks for your contribution. But I have a doubt "how can the user submit the experiment by template?" I saw the REST api `/api/v1/template/submit`, but I can't get its meaning. Such as 'create new template' or others? In my opinion, if we provide the template resource, the experiment resource depends on template. So we should extend the experiment api to support accepts the template id and related params to training.

@jiwq, Thanks for your review

My last PR (SUBMARINE-559)
[API] Define Swagger API for pre-defined template registration/delete, etc.
Support methods such as create new template or delete template
# get template list
/api/v1/template

# get template
/api/v1/template/{templateName}

# post template
/api/v1/template

# patch template
/api/v1/template/{templateName}

# delete template
/api/v1/template/{templateName}

then this PR uses the registered template to submit as an experiment
post /api/v1/template/submit

**Comment by jiwq** [comment]
> > @JohnTing Thanks for your contribution. But I have a doubt "how can the user submit the experiment by template?" I saw the REST api `/api/v1/template/submit`, but I can't get its meaning. Such as 'create new template' or others? In my opinion, if we provide the template resource, the experiment resource depends on template. So we should extend the experiment api to support accepts the template id and related params to training.
> 
> @jiwq, Thanks for your review
> 
> My last PR (SUBMARINE-559)
> [API] Define Swagger API for pre-defined template registration/delete, etc.
> Support methods such as create new template or delete template
> 
> ### get template list
> /api/v1/template
> 
> ### get template
> get /api/v1/template/{templateName}
> 
> ### post template
> post /api/v1/template
> 
> ### patch template
> patch /api/v1/template/{templateName}
> 
> ### delete template
> delete /api/v1/template/{templateName}
> 
> then this PR uses the registered template to submit as an experiment
> 
> ### submit template as an experiment
> post /api/v1/template/submit
> 
> The format of the registered template includes a complete experiment, but the template maker can change some parameters into variables for template users to adjust.

Got. I think `POST /api/v1/template/submit` is not a good design, extend the `POST /api/v1/experiment` is the better selected (more info see mine above comment). Any thoughts?

**Comment by JohnTing** [comment]
@jiwq 
If I only need to change the code place and path, there is no problem.
Do I need to extend the API as POST /api/v1/experiment/submit or other?

**Comment by jiwq** [comment]
> @jiwq
> If I only need to change the code place and path, there is no problem.
> Do I need to extend the API as POST /api/v1/experiment/submit or POST /api/v1/experiment/?

I recommend use `/api/v1/experiment` as the resource and extend the associated operations.

**Comment by xunliu** [comment]
@JohnTing , What is the status of this PR now? Has this PR been completed?

**Review by xunliu — COMMENTED** [review]
@JohnTing , What is the status of this PR now? Has this PR been completed?

**Comment by JohnTing** [comment]
> @JohnTing , What is the status of this PR now? Has this PR been completed?

Thanks for @xunliu comment, this pr has been completed.

**Review by xunliu — APPROVED** [review]
@JohnTing Thank you contibution this feature.
Will merge if no more comments.

**Pull request closed without merging**